src/glossary.adoc

```diff
@@ -10,13 +10,15 @@ A macro can be _fixed arity_ or _variable arity_.
 
 argument::
 A single sub-expression within a macro invocation, corresponding to one of the macro’s
-parameters.  This is by default a one-to-one relation, but a parameter’s grouping mode
-can change this to a many-to-one relation.
+parameters.  This is by default a one-to-one relation, but a rest parameter accepts
+zero-or-more or one-or-more arguments.
 
 argument group::
-The concrete syntax for a _grouped parameter_.
-In a macro invocation, a list containing multiple arguments, delimited explicitly with `[...]`
-notation in Ion text. Used to express parameters that accept more than one argument.
+The concrete syntax for passing a stream of values to a macro parameter.
+In a text E-expression, a group starts with the digraph `(:` and ends with `)`,
+similar to an S-expression.
```

Comment:
Suppose I have macros named "foo" and "bar". How do we prevent this ambiguity?
```
(:foo (:bar baz))
```
Is this an invocation of `foo` with an expression group of `'bar' 'baz'` or is it an invocation of `foo` where the argument is the result from the invocation of `(:bar baz)`?

Reply:
I talked this over with Matt offline. The solution--that whitespace following `:` is the differentiator--feels like a big stumbling block. I think newcomers would often incorrectly consider:
```
(:foo (:bar baz))
```
and
```
(:foo (: bar baz))
```
to be equivalent. To be clear, I think this is just a syntactic issue. I don't object to the underlying concept of expression groups being their own construct.

Reply:
My intention was that whitespace or `)` must follow. I'm also inclined to spec, at least early on, that a macro-ref can _only_ be followed by whitespace or `)`.  This is intended to avoid extending the tokenization weirdness we already have inside S-expressions and make the E-expression "opening" visually separate from its content.

I don't think there's ambiguity; it's in the spec that the characters immediately after `(:` must be a macro reference or `)`, else there's an error; this adds a distinct case for whitespace.  A user who mistakes the difference would likely learn from the immediate syntax error.

I'm not sure why the `[` markers would be more intuitive.  In the examples here I'm surprised by the lack of commas, which adds the need for sexp-like rules regarding tokenization, eg `[:bar@foo]`.

Adding a closing marker introduces another question regarding empty-group;  `[::]` feels off to me.

_Perhaps_ the way to slice this is to change the requirement that the leading colon must abut the paren, and instead tokenize it with the following identifier/integer.  Thus allowing `( :macro ...)`.  Then we can teach that the colon is either (and only) the start of the macro-ref or a group.  But I hesitate because (a) elsewhere in Ion "whitespace" includes commas, and (b) this means the parser may need to scan an unbounded amount of "whitespace" looking for that colon.  This is already a problem with annotations, and I'm loathe to repeat it.

Reply:
> I don't think there's ambiguity; it's in the spec that the characters immediately after (: must be a macro reference or ), else there's an error; this adds a distinct case for whitespace. A user who mistakes the difference would likely learn from the immediate syntax error.

It's not ambiguous for a parser, but it's ambiguous if you're a human. :)

> I'm not sure why the [ markers would be more intuitive. In the examples here I'm surprised by the lack of commas, which adds the need for sexp-like rules regarding tokenization, eg [:bar@foo]

I'm not prescribing a solution, I'm illustrating that having a syntax wherein expressions as similar as `(:foo)` and `(: foo)` having two different meanings is readily avoidable. Users will definitely learn of their mistake when they stub their toe on it, but I think we could spare them that trouble.

> Perhaps the way to slice this is to change the requirement that the leading colon must abut the paren,

I would also prefer not to because of the reasons you stated.

Reply:
> I'm not sure why the `[` markers would be more intuitive.

I'm not looking for "more intuitive"—just something that is less "potential foot-gun" than `(: foo)` vs `(:foo)`.

It just so happens that `[:` is currently not legal Ion, so it's something that _could_ be used.

What I find incredibly *un*intuitive is that these are all equivalent:
* `(foo)` and `( foo)`
* `[foo]` and `[ foo]`
* `{foo:1}` and `{ foo:1}`
* `{{abc12=}}` and `{{ abc12=}}`
 
...but `(:foo)` and `(: foo)` are not equivalent _and_ they are both valid Ion.

> I'm surprised by the lack of commas, which adds the need for sexp-like rules regarding tokenization

If we do use `[:` or similar, then I think commas are fine.

If we're concerned about tokenization simplicity or efficiency, then why not use something like `[:`. Instead of having to check three characters to determine whether we're looking at the start of a macro vs an expression group, we only have to look at two characters—`(: ` vs `[:`. If we know because of the context that it _must_ be either a macro or expression group, then we only actually need to look at the first character—`(` or `[`—to determine which it is.

Reply:
Sorry if I'm belaboring this point, but I thought of one last thing. This will be the last one, I promise.

> Users will definitely learn of their mistake when they stub their toe on it, but I think we could spare them that trouble.

Ideally, we have lots of Ion users (who could each potentially make this mistake), and even if users learn and are aware of it, it's easy to accidentally add a single whitespace. We should assume that users (collectively) will make this mistake over and over, so we should try to minimize the potential impact of and time to identify the mistake. 

When a user accidentally changes e.g. `(:foo "bar-" (:baz a b c))` to `(:foo "bar-" (: baz a b c))`:
* If argument groups start with `[:` (or similar), then they _only_ stub their toe because the result is invalid Ion. A syntax highlighter can tell you immediately that it's syntactically wrong and exactly where it is wrong.
* If argument groups start with `(: `, then the user potentially shoots their foot off by producing new data that is valid Ion but different than what they intended. It _might_ end up failing a cardinality check, but that's not guaranteed. Valid, subtly different data can only reliably be detected by a tool (or person) that is capable of understanding the application-defined semantics of the data. The problem might not be detected right away, and once it is identified that there is a problem, it won't necessarily be obvious where the mistake was made (especially if the "new" data is passed around between multiple systems or transformed in some way before a discrepancy is noticed).

If you use one of `[:`, `[>`, etc. _and_ have commas between elements in the argument group, then it's even more obvious when you have a macro vs an argument group, and it's significantly more difficult to turn an argument group into a macro invocation (or vice versa) by accident.

Reply:
I think it would help me understand your point of view if you can explain why you think `(: ` is the best syntax for argument groups.

Reply:
There's been a lot of good discussion on the syntax, but I'd like to avoid blocking this PR on it.  Changing the concrete syntax with low impact (since there's currently no usage) and happen next, but many other low-level things pivot on moving the grouping concept from parameter-declaration time to invocation-time.  I'd like to unblock that work by getting this merged while leaving the syntax open for discussion.

Reply:
It occurs to me that another potential syntax for argument groups is `(, ...)` since comma, like colon, cannot appear unquoted in an S-expression.  It arguably carries the correct connotation of "sequence of elements", and might be better aligned with the current template-language equivalent `(; ...)`

I would not allow comma inside such a comma-group. I would want everything but the comma to follow S-expression tokenization rules, so that it's consistent with the TL syntax.